File: link_list/lru.cpp
// https://leetcode.com/problems/lru-cache/description/


#include <bits/stdc++.h>
using namespace std;

struct Node {
    int key;
    int value;
    Node* prev = NULL;
    Node* next = NULL;
    Node(int k, int v) : key(k), value(v), prev(nullptr), next(nullptr) {}
};

class LRUCache {
private:
    int capacity;
    // unordered_map<int, pair<int, Node*>> cache; // key -> [value, node
    // pointer]
    unordered_map<int, Node*> cache; // key -> node pointer
    Node* head;                      // dummy head
    Node* tail;                      // dummy tail

    // Node* addNode(int key, int value) {
    //     Node* node = new Node(key, value);
    void addNode(Node* node) {
        node->next = head;
        node->prev = head->prev;

        head->prev->next = node;
        head->prev = node;

        // return node;
    }

    void removeNode(Node* node) {
        node->next->prev = node->prev;
        node->prev->next = node->next;
        // delete node;
    }

    // Move existing node to front
    void moveToFront(Node* node) {
        removeNode(node);
        addNode(node);
    }

public:
    LRUCache(int capacity) {
        this->capacity = capacity;
        head = new Node(-1, -1);
        tail = new Node(-1, -1);

        tail->next = head;
        head->prev = tail;
    }

    int get(int key) {

        if (cache.find(key) == cache.end()) {
            return -1;
        }
        Node* node = cache[key];
        moveToFront(node);

        return node->value;
    }

    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            Node* node = cache[key];
            // cache.erase(key);
            // removeNode(node);
            node->value = value;
            moveToFront(node);
            return;
        } else if (cache.size() == capacity) {
            Node* last = tail->next;
            cache.erase(last->key);
            removeNode(last);
            delete last;
        }

        // Node* node = addNode(key, value);
        // cache[key] = {value, node};
        Node* node = new Node(key, value);
        addNode(node);
        cache[key] = node;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: js-codes/lru.ts
// https://leetcode.com/problems/lru-cache/description/

class LLNode {
  key: number;
  value: number;
  prev: LLNode | null;
  next: LLNode | null;
  constructor(key: number, value: number) {
    this.key = key;
    this.value = value;
    this.prev = null;
    this.next = null;
  }
}

class DLList {
  head: LLNode | null;
  tail: LLNode | null;
  constructor() {
    this.head = new LLNode(-1, -1);
    this.tail = new LLNode(-1, -1);
    this.head.next = this.tail;
    this.tail.prev = this.head;
  }
  addNode(node: LLNode) {
    this.tail!.prev!.next! = node;
    node.prev = this.tail!.prev!
    node.next = this.tail
    this.tail!.prev! = node
  }
  moveNode(node: LLNode) {
    // remove node
    node.prev!.next! = node.next!
    node.next!.prev! = node.prev!;

    // move node to tail
    this.addNode(node)
  }
}

class LRUCache {
  dll: DLList;
  mapToNode: Map<number, LLNode>;
  capacity: number;
  constructor(capacity: number) {
    this.dll = new DLList();
    this.mapToNode = new Map<number, LLNode>();
    this.capacity = capacity;
  }

  get(key: number): number {
    const node = this.mapToNode.get(key)
    if (!node) return -1;
    this.dll.moveNode(node)
    return node.value!;
  }

  put(key: number, value: number): void {
    if(this.mapToNode.get(key)){
        const node = this.mapToNode.get(key)!
        node.value = value;
        this.dll.moveNode(node)
    }
    else{
        if(this.capacity === this.mapToNode.size){
            const firstNode = this.dll.head?.next!
            this.dll.moveNode(firstNode)
            this.mapToNode.delete(firstNode.key)
            firstNode.key = key
            firstNode.value = value
            this.mapToNode.set(key, firstNode)
        }
        else{
            const node = new LLNode(key, value);
            this.mapToNode.set(key, node)
            this.dll.addNode(node)
        }
    }
  }
}
